File: 2 kyu/Regular expression parser/solution.hs
module RegExpParser (RegExp(..), parseRegExp) where

import Control.Applicative (pure, empty, (<|>), (<*>), (<$>) , Applicative, Alternative)
import Control.Monad (ap, (>=>), guard)
import Control.Arrow ((&&&))

data RegExp = Normal Char | Any | ZeroOrMore RegExp | Or RegExp RegExp | Str [RegExp] deriving (Show, Eq)

newtype Parser a = Parser {runParser :: String -> Maybe (a, String)}

instance Monad Parser where
    return x = Parser (\s -> Just (x, s))
    (Parser f) >>= g = Parser (f >=> (\(x, s) -> runParser (g x) s))

instance Functor Parser where
    fmap f = ap (return f)

instance Applicative Parser where
    pure = return
    fa <*> xa = fa >>= \f -> xa >>= \x -> return (f x)

instance Alternative Parser where
    empty = Parser (const Nothing)
    (Parser f) <|> (Parser g) = Parser (uncurry (<|>) . (f &&& g))
    

liftToken :: (Char -> Maybe a) -> Parser a
liftToken f = Parser g
    where g [] = Nothing
          g (c:cs) = f c >>= (\x -> Just (x, cs))

eat :: Char -> Parser ()
eat t = liftToken (guard . (==t))

wrap :: Char -> Parser a -> Char -> Parser a
wrap l p r = eat l >> 
             p >>= 
             \x -> eat r >>
             return x 
            
parseList :: Parser a -> Parser [a]
parseList itemP = listP
    where listP = ((:) <$> itemP <*> listP) <|> return []

(~>) :: Char -> a -> Parser a
c ~> x = eat c >> return x

atomP :: Parser RegExp
atomP = liftToken isChar <|> ('.' ~> Any) <|> wrap '(' regExpP ')'
  where isChar c | c `elem` ".*|()" = Nothing
                 | otherwise        = Just (Normal c)

zeroOrMoreP :: Parser RegExp
zeroOrMoreP = (atomP >>= 
               \a -> eat '*' >> 
               return (ZeroOrMore a)) <|> atomP

seqP :: Parser RegExp
seqP = (zeroOrMoreP >>= 
        \a -> zeroOrMoreP >>= 
        \b -> parseList zeroOrMoreP >>= 
        \c -> return (Str (a:b:c))) <|> zeroOrMoreP
    
regExpP :: Parser RegExp
regExpP = (seqP >>=
           \l -> eat '|' >>
           seqP >>= 
           \r -> return (Or l r)) <|> seqP
    
parseRegExp :: String -> Maybe RegExp
parseRegExp s = case runParser regExpP s of Just (x, "") -> Just x
                                            _            -> Nothing
________________________________________
module RegExpParser
       ( RegExp(..)
       , parseRegExp
       ) where

import Text.ParserCombinators.ReadP
import Control.Monad
import Control.Applicative ((<$>),(<*>),(<*),(*>))

data RegExp = Normal Char       -- ^ A character that is not in "()*|."
            | Any               -- ^ Any character
            | ZeroOrMore RegExp -- ^ Zero or more occurances of the same regexp
            | Or RegExp RegExp  -- ^ A choice between 2 regexps
            | Str [RegExp]      -- ^ A sequence of regexps.
  deriving (Show, Eq)


parseRegExp :: String -> Maybe RegExp
parseRegExp s = helper $ readP_to_S regExp s
  where
    helper rs = case [a | (a,"") <- rs] of
      []    -> Nothing
      (a:_) -> Just a

regExp :: ReadP RegExp
regExp = strExp +++ (Or <$> (strExp <* char '|') <*> strExp)

strExp :: ReadP RegExp
strExp = charExp
     +++ (Str <$> many1 charExp)

charExp :: ReadP RegExp
charExp = atom +++ (ZeroOrMore <$> atom <* char '*')

atom :: ReadP RegExp
atom = (Normal <$> satisfy (not . flip elem "()*|."))
   +++ (char '.' *> return Any)
   +++ (char '(' *> regExp <* char ')')
________________________________________
module RegExpParser
( RegExp(..)
, parseRegExp
) where

    import Control.Applicative
    import Data.Char

    newtype Parser a = Parser (String -> Maybe (a, String))
    
    parse :: Parser a -> String -> Maybe (a, String)
    parse (Parser parser) = parser

    item :: Parser Char
    item = Parser (\ st -> case st of 
                            [] -> Nothing
                            (x : xs) -> Just (x , xs))

    instance Functor Parser where
        -- fmap :: (a -> b) -> Parser a -> Parser b
        fmap f p = Parser (\ st -> case parse p st of
                                        Nothing -> Nothing
                                        Just (v, out) -> Just (f v, out))

    instance Applicative Parser where

        -- pure :: a -> Parser a
        pure v = Parser (\ st -> Just (v, st))

        -- (<*>) :: Parser (a -> b) -> Parser a -> Parser b
        pab <*> pa = Parser (\ st -> case parse pab st of
            Nothing -> Nothing
            Just (v, out) -> parse (fmap v pa) out)

    instance Monad Parser where

        -- (>>=) :: Parser a -> (a -> Parser b) -> Parser b
        pa >>= apb = Parser (\ st -> case parse pa st of
            Nothing -> Nothing
            Just (a, out) -> parse (apb a) out)

    instance Alternative Parser where

        -- empty :: Parser a
        empty = Parser $ const Nothing

        -- <|> :: Parser a -> Parser a -> Parser a
        pa <|> pa2 = Parser (\ st -> case parse pa st of
            Nothing -> parse pa2 st
            Just x -> Just x)

    data RegExp = Normal Char       -- ^ A character that is not in "()*|."
        | Any               -- ^ Any charater
        | ZeroOrMore RegExp -- ^ Zero or more occurances of the same regexp
        | Or RegExp RegExp  -- ^ A choice between 2 regexps
        | Str [RegExp]      -- ^ A sequence of regexps.
        deriving (Show, Eq)

    failure :: Parser a
    failure = Parser $ const Nothing

    sat :: (Char -> Bool) -> Parser Char
    sat f = do 
        x <- item
        if f x then return x else failure

    digit :: Parser Char
    digit = sat isDigit

    lower :: Parser Char
    lower = sat isLower

    upper :: Parser Char
    upper = sat isUpper

    letter :: Parser Char
    letter = sat isLetter

    alphanum :: Parser Char
    alphanum = sat isAlphaNum

    char :: Char -> Parser Char
    char x = sat (==x)

    oneOf :: String -> Parser Char
    oneOf "" = failure
    oneOf [x] = char x
    oneOf (x : xs) = char x <|> oneOf xs

    noneOf :: String -> Parser Char
    noneOf "" = item
    noneOf str = Parser (\ st -> if (st == "") || head st `elem` str then Nothing
        else Just (head st, tail st))

    parseSimple :: Parser RegExp
    parseSimple = do
        c <- noneOf "()*|."
        return $ Normal c
        <|> do
            _ <- char '.'
            return Any
            <|> do
                _ <- char '('
                x <- parseOr
                _ <- char ')'
                return x

    parseStr :: Parser RegExp
    parseStr = do
        x <- parseMulti
        xs <- some parseMulti
        return $ Str (x : xs)
        <|> parseMulti

    parseMulti :: Parser RegExp
    parseMulti = do
        x <- parseSimple
        _ <- char '*'
        return $ ZeroOrMore x
        <|> parseSimple

    parseOr :: Parser RegExp
    parseOr = do
        x <- parseStr
        _ <- char '|'
        y <- parseStr
        return $ Or x y
        <|> parseStr

    parseRegExp :: String -> Maybe RegExp
    parseRegExp s = case parse parseOr s of
        Just (x , []) -> Just x
        _ -> Nothing
        
________________________________________
module RegExpParser
  ( RegExp (..),
    parseRegExp,
  )
where

import Data.Functor (($>))
import Text.Parsec
import Text.Parsec.String

data RegExp
  = -- | A character that is not in "()*|."
    Normal Char
  | -- | Any charater
    Any
  | -- | Zero or more occurances of the same regexp
    ZeroOrMore RegExp
  | -- | A choice between 2 regexps
    Or RegExp RegExp
  | -- | A sequence of regexps.
    Str [RegExp]
  deriving (Show, Eq)

res = "()*|."

pAny :: Parser RegExp
pAny = char '.' $> Any

pNormal :: Parser RegExp
pNormal = Normal <$> noneOf res

parens = between (char '(') (char ')')

pFactor = parens pExpr <|> pNormal <|> pAny

pExpr = chainl1 (toStr <$> many1 pTerm) (char '|' $> Or)

pTerm = do
  f <- pFactor
  p <- try (char '*' $> ZeroOrMore) <|> return id
  return $ p f

toStr [x] = x
toStr xs = Str xs

parseRegExp :: String -> Maybe RegExp
parseRegExp s = case parse (pExpr <* eof) "RegExp" s of
  Left _ -> Nothing
  Right x -> Just x
________________________________________
module RegExpParser
       ( RegExp(..)
       , parseRegExp
       ) where
       
import Text.ParserCombinators.ReadP
import Data.List (nub)

data RegExp = Normal Char       -- ^ A character that is not in "()*|."
            | Any               -- ^ Any charater
            | ZeroOrMore RegExp -- ^ Zero or more occurances of the same regexp
            | Or RegExp RegExp  -- ^ A choice between 2 regexps
            | Str [RegExp]      -- ^ A sequence of regexps.
  deriving (Show, Eq)

metacharacters = "()*|."

char' :: ReadP RegExp
char' = Normal <$> satisfy (`notElem` metacharacters)

any' :: ReadP RegExp
any' = char '.' >> return Any

group :: ReadP RegExp
group = do char '('
           e <- expression
           char ')'
           return e

atom :: ReadP RegExp
atom = group +++ any' +++ char'

star :: ReadP RegExp
star = do e <- atom
          char '*'
          return $ ZeroOrMore e

factor :: ReadP RegExp
factor = star <++ atom

manyFactor :: ReadP RegExp
manyFactor = do e <- factor
                es <- many1 factor
                return $ Str (e:es)

term :: ReadP RegExp
term = star +++ (manyFactor) +++ factor

union :: ReadP RegExp
union = do e1 <- term
           char '|'
           e2 <- expression
           return (Or e1 e2)

expression :: ReadP RegExp
expression = union +++ term


parseRegExp :: String -> Maybe RegExp
parseRegExp s = case readP_to_S expression s of
                  [] -> Nothing
                  xs -> case nub $ filter ((==) "" . snd) xs of
                    [(a, _)] -> Just a
                    _ -> Nothing
________________________________________
module RegExpParser
       ( RegExp(..)
       , parseRegExp
       ) where

data RegExp = Normal Char       -- ^ A character that is not in "()*|."
            | Any               -- ^ Any charater
            | ZeroOrMore RegExp -- ^ Zero or more occurances of the same regexp
            | Or RegExp RegExp  -- ^ A choice between 2 regexps
            | Str [RegExp]      -- ^ A sequence of regexps.
  deriving (Show, Eq)

getRegExpList :: RegExp -> Maybe [RegExp]
getRegExpList (Str sx) = Just sx
getRegExpList _ = Nothing

data Token = TChar Char
           | TBracketBegin
           | TBracketEnd
           | TAny
           | TOr
           | TZeroOrMore
           | TEOF

parseRegExp :: String -> Maybe RegExp
parseRegExp = grammarize . tokenize

data TokenKind = KChar
               | KBracketBegin
               | KBracketEnd
               | KAny
               | KOr
               | KZeroOrMore
               | KEOF
               deriving (Eq)

kind :: Token -> TokenKind
kind (TChar _)       = KChar
kind (TBracketBegin) = KBracketBegin
kind (TBracketEnd)   = KBracketEnd
kind (TAny)          = KAny
kind (TOr)           = KOr
kind (TZeroOrMore)   = KZeroOrMore
kind (TEOF)          = KEOF

isKind :: Token -> TokenKind -> Bool
isKind = (==) . kind

-- Lexical

-- TChar         ::= ~["(",")",".","|","*"]
-- TBracketBegin ::= "("
-- TBracketEnd   ::= ")"
-- TAny          ::= "."
-- TOr           ::= "|"
-- TZeroOrMore   ::= "*"
-- TEOF          ::= ""

tokenize :: String -> [Token]
tokenize [] = [TEOF]
tokenize (x:sx) = (case x of '(' -> TBracketBegin
                             ')' -> TBracketEnd
                             '.' -> TAny
                             '|' -> TOr
                             '*' -> TZeroOrMore
                             _ -> TChar x) : tokenize sx
-- Grammar

-- TopLevel       ::= RegExpList() (Or())? < EOF >
-- Or             ::= < TOr > RegExpList()
-- RegExpList     ::= RegExp() (RegExpListTail())?
-- RegExpListTail ::= (RegExp())+
-- RegExp         ::= (Normal() | Any() | CapturingGroup()) (ZeroOrMore())?
-- Normal         ::= < TChar >
-- Any            ::= < TAny >
-- ZeroOrMore     ::= < TZeroOrMore >
-- CapturingGroup ::= < TBracketBegin > RegExpList() (Or())? < TBracketEnd >

grammarize :: [Token] -> Maybe RegExp
grammarize = snd . compileTopLevel

checkRegExp :: ([Token], Maybe RegExp) -> ([Token], Maybe RegExp) -> ([Token], Maybe RegExp)
checkRegExp previous@(_, Nothing) next = previous
checkRegExp previous@(_, Just _) next = next

compileTopLevel :: [Token] -> ([Token], Maybe RegExp)
compileTopLevel tokens = checkRegExp result $ case kind $ head tokens' of {
    KOr  -> let {
      result'@(tokens''', rightRegExpList) = compileOr tokens';
      tokens'''' = tail tokens''';
    } in checkRegExp result' $ case kind $ head tokens''' of {
        KEOF -> (tokens'''', pure Or <*> leftRegExpList <*> rightRegExpList);
        _    -> (tokens''', Nothing);
      };
    KEOF -> (tokens'', leftRegExpList);
    _    -> (tokens', Nothing);
  }
  where
    result@(tokens', leftRegExpList) = compileRegExpList tokens
    tokens'' = tail tokens'

compileOr :: [Token] -> ([Token], Maybe RegExp)
compileOr tokens = case kind $ head tokens of {
    KOr -> compileRegExpList tokens';
    _   -> (tokens, Nothing);
  }
  where
    tokens' = tail tokens

compileRegExpList :: [Token] -> ([Token], Maybe RegExp)
compileRegExpList tokens = checkRegExp result $ case kind $ head tokens' of {
    x | (x == KChar)         ||
        (x == KBracketBegin) ||
        (x == KAny)             -> let {
          result'@(tokens'', regExpListTail) = compileRegExpListTail tokens'
        } in checkRegExp result' $ (tokens'', pure Str <*> (pure (:) <*> regExp <*> (regExpListTail >>= getRegExpList)));
    _                           -> (tokens', regExp)
  }
  where
    result@(tokens', regExp) = compileRegExp tokens

compileRegExpListTail :: [Token] -> ([Token], Maybe RegExp)
compileRegExpListTail tokens = checkRegExp result $ case kind $ head tokens' of {
    x | (x == KChar)         ||
        (x == KBracketBegin) ||
        (x == KAny)             -> let {
          result'@(tokens'', regExpListTail) = compileRegExpListTail tokens'
        } in checkRegExp result' $ (tokens'', pure Str <*> (pure (:) <*> regExp <*> (regExpListTail >>= getRegExpList)));
    _                           -> (tokens', pure (Str . (:[])) <*> regExp)
  }
  where
    result@(tokens', regExp) = compileRegExp tokens

compileRegExp :: [Token] -> ([Token], Maybe RegExp)
compileRegExp tokens = let {
    result@(tokens', regExp) = case kind $ head tokens of {
      KChar         -> compileNormal tokens;
      KAny          -> compileAny tokens;
      KBracketBegin -> compileCapturingGroup tokens;
      _             -> (tokens, Nothing);
    };
    tokens'' = tail tokens'
  } in checkRegExp result $ case kind $ head tokens' of {
      KZeroOrMore -> (tokens'', pure ZeroOrMore <*> regExp);
      _           -> (tokens', regExp);
    }

compileNormal :: [Token] -> ([Token], Maybe RegExp)
compileNormal tokens = case head tokens of {
    TChar c -> (tail tokens, Just $ Normal c);
    _       -> (tokens, Nothing);
  }

compileAny :: [Token] -> ([Token], Maybe RegExp)
compileAny tokens = case kind $ head tokens of {
    KAny -> (tail tokens, Just Any);
    _    -> (tokens, Nothing);
  }

compileCapturingGroup :: [Token] -> ([Token], Maybe RegExp)
compileCapturingGroup tokens = case kind $ head tokens of {
    KBracketBegin -> let {
      result@(tokens'', leftRegExpList) = compileRegExpList tokens'
    } in checkRegExp result $ let {
        result''@(tokens'''', capturingGroup) = case kind $ head tokens'' of {
          KOr -> let {
            result'@(tokens''', rightRegExpList) = compileOr tokens''
          } in (tokens''', pure Or <*> leftRegExpList <*> rightRegExpList);
          _   -> (tokens'', leftRegExpList);
        };
        tokens''''' = tail tokens'''';
      } in checkRegExp result'' $ case kind $ head tokens'''' of {
          KBracketEnd -> (tokens''''', capturingGroup);
          _           -> (tokens'''', Nothing);
        };
    _             -> (tokens, Nothing);
  }
  where
    tokens' = tail tokens

File: 2 kyu/Regular expression parser/solution.c
5470c635304c127cad000f0d


typedef struct RegExp RegExp;

RegExp* any ();
RegExp* normal (char c);
RegExp* zeroOrMore (RegExp *starred);
RegExp* or (RegExp *left, RegExp *right);
RegExp* str (RegExp *first);
RegExp* add (RegExp *str, RegExp *next);


RegExp *parseRegExp (char *input);
RegExp *parse_or(char *input, int *i);
RegExp *parse_str(char *input, int *i);
RegExp *parse_zero_or_more(char *input, int *i);
RegExp *parse_group(char *input, int *i);
RegExp *parse_any(char *input, int *i);
RegExp *parse_normal(char *input, int *i);


RegExp *parseRegExp (char *input) {
  int i = 0;
  RegExp *ast = parse_or(input, &i);
  if (i != strlen(input)) return 0;
  return ast;
}

RegExp *parse_or(char *input, int *i) {
  RegExp *next = parse_str(input, i);
  if (input[*i] == '|') {
    (*i)++;
    return or(next, parse_str(input, i));
  } else {
    return next;
  }
};

RegExp *parse_str(char *input, int *i) {
  RegExp *r = parse_zero_or_more(input, i);
  RegExp *n = parse_zero_or_more(input, i);
  if (n) {
    r = str(r);
    add(r, n);
    while ( n = parse_zero_or_more(input, i) ) add(r, n);
  }
  return r;
};


RegExp *parse_zero_or_more(char *input, int *i) {
  RegExp *next = parse_group(input, i);
  if (next && input[*i] == '*') {
    (*i)++;
    return zeroOrMore(next);
  }
  return next;
};


RegExp *parse_group(char *input, int *i) {
  if (input[*i] == '(') {
    (*i)++;
    RegExp *next = parse_or(input, i);
    if (input[*i] != ')') return 0;
    (*i)++;
    return next;
  } else {
    return parse_any(input, i);
  }
};

RegExp *parse_any(char *input, int *i) {
  return input[*i] == '.' ? ((*i)++, any()) : parse_normal(input, i);
}

RegExp *parse_normal(char *input, int *i) {
  switch (input[*i]) {
    case '*':
    case '|':
    case '(':
    case ')':
    case '\0':
      return 0;
    default:
      return normal(input[(*i)++]);
  }
}
____________________________________________________________
typedef struct RegExp RegExp;

RegExp* any ();
RegExp* normal (char c);
RegExp* zeroOrMore (RegExp *starred);
RegExp* or (RegExp *left, RegExp *right);
RegExp* str (RegExp *first);
RegExp* add (RegExp *str, RegExp *next);

typedef struct state { char *s; void *r; } state;

state  oneOf (char *input, char *s) { return *input &&  strchr (s, *input) ? (state) {input+1, (void*)*input } : (state) {input}; }
state noneOf (char *input, char *s) { return *input && !strchr (s, *input) ? (state) {input+1, (void*)*input } : (state) {input}; }
state parseExpr (char *input); state parseSeq (char *input);
state parseNormal (char *input) { state s = noneOf (input, "()*|."); if (s.r) s.r = normal (*input); return s; }
state parseAny  (char *input) { state s = oneOf (input, "."); if (s.r) s.r = any (); return s; }
state parseMany (char *input) { state a,s = parseExpr (input); if (s.r) a = oneOf (s.s, "*"); return s.r && a.r ? (state){a.s, zeroOrMore (s.r)} : s; }
state parseOr (char *input) { state a,s = parseSeq (input); return (a = oneOf (s.s, "|")).r ? a = parseSeq (a.s), (state){a.s, or (s.r, a.r)} : s; }
RegExp *parseRegExp (char *input) { state s = parseOr (input); return *s.s ? 0 : s.r; }

state parseExpr (char *input) {
  RegExp *r; state s = parseNormal (input);
  if (!s.r) s = parseAny (s.s);
  if (!s.r) { s = oneOf (s.s, "("); if (s.r) s = parseOr (s.s); if (r=s.r) s = oneOf (s.s, ")"); s.r = s.r ? r : 0; }
  return s;
}

state parseSeq (char *input) {
  state s = parseMany (input), a = parseMany (s.s); RegExp *r = str (s.r);
  if (a.r) { while (a.r) { add (r, a.r); a = parseMany (a.s); } s = (state) {a.s, r}; }
  return s;
}
____________________________________________________________
#include <stdbool.h>
#include <stdlib.h>

typedef struct RegExp RegExp;

RegExp* any ();
RegExp* normal (char c);
RegExp* zeroOrMore (RegExp *starred);
RegExp* or (RegExp *left, RegExp *right);
RegExp* str (RegExp *first);
RegExp* add (RegExp *str, RegExp *next);

// Grammar:
//
//   RE       -> Term RETail
//   RETail   -> "|" Term | empty
//   Term     -> Factor TermTail
//   TermTail -> Term | empty
//   Factor   -> Item Stars
//   Item     -> atom | "(" RE ")"    atom is either a "normal" character or ".".
//   Stars    -> "*" | empty
//
// First sets:
//
//   RE       : atom, "("
//   RETail   : "|", empty
//   Term     : atom, "("
//   TermTail : atom, "(", empty
//   Factor   : atom, "("
//   Item     : atom, "("
//   Stars    : "*", empty
//
// Follow sets:
//
//   RE       : ")", empty
//   RETail   : ")", empty
//   Term     : ")", "|", empty
//   TermTail : ")", "|", empty
//   Factor   : atom, "(", ")", "|", empty
//   Item     : atom, "(", ")", "*", "|", empty
//   Stars    : atom, "(", ")", "|", empty
//
// Predict sets:
//
//   RE       -> Term RETail     : atom, "("
//   RETail   -> "|" Term        : "|"
//   RETail   -> empty           : ")", empty
//   Term     -> Factor TermTail : atom, "("
//   TermTail -> Term            : atom, "("
//   TermTail -> empty           : ")", "|", empty
//   Factor   -> Item Stars      : atom, "("
//   Item     -> atom            : atom
//   Item     -> "(" RE ")"      : "("
//   Stars    -> "*" Stars       : "*"
//   Stars    -> empty           : atom, "(", ")", "|", empty

static RegExp *parseRE(const char **input);
static RegExp *parseRETail(const char **input, RegExp *head);
static RegExp *parseTerm(const char **input);
static RegExp *parseTermTail(const char **input, RegExp *term, bool first);
static RegExp *parseFactor(const char **input);
static RegExp *parseItem(const char **input);
static RegExp *parseStars(const char **input, RegExp *item);

RegExp *parseRE(const char **const input) {
  switch (**input) {
  case ')':
  case '*':
  case '|':
  case '\0':
    // Syntax error.
    return NULL;
  
  default: { // atom, "("
    // RE -> Term RETail
    
    RegExp *const ast = parseTerm(input);
    if (!ast) return NULL;
    
    return parseRETail(input, ast);
  }
  }
}

RegExp *parseRETail(const char **const input, RegExp *const head) {
  switch (**input) {
  case '|':
    // RETail -> "|" Term
    
    (*input)++;
    
    RegExp *const ast = parseTerm(input);
    if (!ast) return NULL;
      
    return or(head, ast);
      
  case ')':
  case '\0':
    // RETail -> empty
    return head;
      
  default:
    // Syntax error.
    return NULL;
  }
}

RegExp *parseTermAux(const char **const input, RegExp *const term, const bool first) {
  switch (**input) {
  case ')':
  case '*':
  case '|':
  case '\0':
    // Syntax error.
    return NULL;
      
  default: { // atom, "("
    // Term -> Factor TermTail
    
    RegExp *const ast = parseFactor(input);
    if (!ast) return NULL;
    
    return parseTermTail(input, term ? add(term, ast) : ast, first);
  }
  }
}

RegExp *parseTerm(const char **const input) {
  return parseTermAux(input, NULL, true);
}

RegExp *parseTermTail(const char **const input, RegExp *const term, const bool first) {
  switch (**input) {
  case ')':
  case '|':
  case '\0':
    // TermTail -> empty
    return term;
  
  case '*':
    // Syntax error.
    return NULL;
  
  default: // atom, "("
    // TermTail -> Term
    return parseTermAux(input, first ? str(term) : term, false);
  }
}

RegExp *parseFactor(const char **const input) {
  switch (**input) {
  case ')':
  case '*':
  case '|':
  case '\0':
    // Syntax error.
    return NULL;
      
  default: { // atom, "("
    // Factor -> Item Stars
    
    RegExp *const ast = parseItem(input);
    if (!ast) return NULL;
    
    return parseStars(input, ast);
  }
  }
}

RegExp *parseItem(const char **const input) {
  switch (**input) {
  case '(':
    // Item -> "(" RE ")"
    
    (*input)++;
      
    RegExp *const ast = parseRE(input);
    if (!ast) return NULL;
      
    if (*(*input)++ != ')') return NULL;
    
    return ast;
  
  case ')':
  case '*':
  case '|':
  case '\0':
    // Syntax error.
    return NULL;
      
  case '.': // atom
    // Item -> atom
    (*input)++;
    return any();
      
  default: // atom
    // Item -> atom
    return normal(*(*input)++);
  }
}

RegExp *parseStars(const char **const input, RegExp *const item) {
  if (**input == '*') {
    // Stars -> "*"
    (*input)++;
    return zeroOrMore(item);
  } else {
    // Stars -> empty
    return item;
  }
}

RegExp *parseRegExp (char *input) {
  RegExp *const ast = parseRE((const char **)&input);
  return *input == '\0' ? ast : NULL;
}
____________________________________________________________
#include <stdlib.h>
#include <stdbool.h>

typedef struct RegExp RegExp;

RegExp* any ();
RegExp* normal (char c);
RegExp* zeroOrMore (RegExp *starred);
RegExp* or (RegExp *left, RegExp *right);
RegExp* str (RegExp *first);
RegExp* add (RegExp *str, RegExp *next);

typedef struct {
  char *input;
} state;

static RegExp *parse(state *s, bool canOr);

RegExp *parseRegExp (char *input) {
  state s;
  s.input = input;
  RegExp *result = parse(&s, true);
  return result;
}

static RegExp *parse(state *s, bool canOr) {
  RegExp *l0 = NULL, *l1 = NULL, *right;
  bool wasStar = false;
  for (;;) {
    if (*s->input == ')')
      break;
    char c = *s->input++;
    if (c == 0)
      break;

    if (c == '*') {
      if (l1 == NULL) return NULL;
      if (wasStar) return NULL;
      l1 = zeroOrMore(l1);
      wasStar = true;
      continue;
    }
    wasStar = false;
    
    if (c == '|') {
      if (!canOr)
        return NULL;

      if (l0)
        l1 = add(l0, l1);
      else if (!l1)
        return NULL;

      right = parse(s, false);
      if (!right)
        return NULL;

      return or(l1, right);
    }

    if (c == '.')
      right = any();
    else if (c == '(') {
      right = parse(s, true);
      if (!right || *s->input++ != ')')
        return NULL;
    } else {
      right = normal(c);
    }

    if (l0)
      l0 = add(l0, l1);
    else if (l1)
      l0 = str(l1);
    l1 = right;
  }

  if (l0)
    return add(l0, l1);
  else
    return l1;
}
____________________________________________________________
#include <stdint.h>
#include <string.h>

typedef struct RegExp RegExp;

RegExp* any ();
RegExp* normal (char c);
RegExp* zeroOrMore (RegExp *starred);
RegExp* or (RegExp *left, RegExp *right);
RegExp* str (RegExp *first);
RegExp* add (RegExp *str, RegExp *next);

typedef struct state { char *s; void *r; } state;

state  oneOf (char *input, char *s) { return *input &&  strchr (s, *input) ? (state) {input+1, (void*)(uintptr_t)*input } : (state) {input, NULL}; }
state noneOf (char *input, char *s) { return *input && !strchr (s, *input) ? (state) {input+1, (void*)(uintptr_t)*input } : (state) {input, NULL}; }
state parseExpr (char *input); state parseSeq (char *input);
state parseNormal (char *input) { state s = noneOf (input, "()*|."); if (s.r) s.r = normal (*input); return s; }
state parseAny  (char *input) { state s = oneOf (input, "."); if (s.r) s.r = any (); return s; }
state parseMany (char *input) { state a,s = parseExpr (input); if (s.r) a = oneOf (s.s, "*"); return s.r && a.r ? (state){a.s, zeroOrMore (s.r)} : s; }
state parseOr (char *input) { state a,s = parseSeq (input); return (a = oneOf (s.s, "|")).r ? a = parseSeq (a.s), (state){a.s, or (s.r, a.r)} : s; }
RegExp *parseRegExp (char *input) { state s = parseOr (input); return *s.s ? 0 : s.r; }

state parseExpr (char *input) {
  RegExp *r; state s = parseNormal (input);
  if (!s.r) s = parseAny (s.s);
  if (!s.r) { s = oneOf (s.s, "("); if (s.r) s = parseOr (s.s); if ((r=s.r)) s = oneOf (s.s, ")"); s.r = s.r ? r : 0; }
  return s;
}

state parseSeq (char *input) {
  state s = parseMany (input), a = parseMany (s.s); RegExp *r = str (s.r);
  if (a.r) { while (a.r) { add (r, a.r); a = parseMany (a.s); } s = (state) {a.s, r}; }
  return s;
}
